### PR TITLE
fix: try finding a socket, otherwise fail, respect user choice

### LINKED ISSUE #1658: `Error: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?` on macOS while docker commands are working
### Bug report info

```plain text
act version:            0.2.43
GOOS:                   darwin
GOARCH:                 amd64
NumCPU:                 8
Docker host:            DOCKER_HOST environment variable is unset/empty.
Sockets found:
	/Users/lena/.docker/run/docker.sock
Config files:           
	/Users/lena/.actrc:
		-P ubuntu-latest=catthehacker/ubuntu:act-latest
		-P ubuntu-22.04=catthehacker/ubuntu:act-22.04
		-P ubuntu-20.04=catthehacker/ubuntu:act-20.04
		-P ubuntu-18.04=catthehacker/ubuntu:act-18.04
Build info:
	Go version:            go1.20.1
	Module path:           command-line-arguments
	Main version:          
	Main path:             
	Main checksum:         
	Build settings:
		-buildmode:           exe
		-compiler:            gc
		-ldflags:             -X main.version=0.2.43
		CGO_ENABLED:          1
		CGO_CFLAGS:           
		CGO_CPPFLAGS:         
		CGO_CXXFLAGS:         
		CGO_LDFLAGS:          
		GOARCH:               amd64
		GOOS:                 darwin
		GOAMD64:              v1

Error: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

### Command used with act

```sh
act
```

### Describe issue

I expected it to not print `Error: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?`, because commands like `docker ps` do work:
```
CONTAINER ID   IMAGE                    COMMAND                  CREATED         STATUS         PORTS                NAMES
b55271169ff4   docker/getting-started   "/docker-entrypoint.…"   2 minutes ago   Up 2 minutes   0.0.0.0:80->80/tcp   funny_euclid
```

### Link to GitHub repository

https://github.com/LenaWil/Squire/commit/d038cc03cd0b075071a2ab001ed226c1b3953225

### Workflow content

```yml
name: Lint Black (official)

# Without review-dog… maybe it’s a bit of waste to make them run twice, but anyway.
on: [push, pull_request]

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: psf/black@stable
```

### Relevant log output

```sh
DEBU[0000] Loading environment from /Users/lena/Documents/Knights/Squire/.env 
DEBU[0000] Loading action inputs from /Users/lena/Documents/Knights/Squire/.input 
DEBU[0000] Loading secrets from /Users/lena/Documents/Knights/Squire/.secrets 
DEBU[0000] Loading workflows from '/Users/lena/Documents/Knights/Squire/.github/workflows' 
DEBU[0000] Loading workflows recursively                
DEBU[0000] Found workflow 'black-review-dog.yml' in '/Users/lena/Documents/Knights/Squire/.github/workflows/black-review-dog.yml' 
DEBU[0000] Found workflow 'black.yml' in '/Users/lena/Documents/Knights/Squire/.github/workflows/black.yml' 
DEBU[0000] Found workflow 'run_tests.yml' in '/Users/lena/Documents/Knights/Squire/.github/workflows/run_tests.yml' 
DEBU[0000] Reading workflow '/Users/lena/Documents/Knights/Squire/.github/workflows/black-review-dog.yml' 
DEBU[0000] Conditional GET for notices etag=89b55634-e474-4cff-a075-2c41abcfb8c3 
DEBU[0000] Reading workflow '/Users/lena/Documents/Knights/Squire/.github/workflows/black.yml' 
DEBU[0000] Reading workflow '/Users/lena/Documents/Knights/Squire/.github/workflows/run_tests.yml' 
DEBU[0000] Preparing plan with all jobs                 
DEBU[0000] Using the only detected workflow event: pull_request 
DEBU[0000] Planning jobs for event: pull_request        
DEBU[0000] Found revision: d038cc03cd0b075071a2ab001ed226c1b3953225 
DEBU[0000] Found revision: d038cc03cd0b075071a2ab001ed226c1b3953225 
DEBU[0000] Found revision: d038cc03cd0b075071a2ab001ed226c1b3953225 
DEBU[0000] Found revision: d038cc03cd0b075071a2ab001ed226c1b3953225 
DEBU[0000]                                              
DEBU[0000] Found revision: d038cc03cd0b075071a2ab001ed226c1b3953225 
ERRO[0000] failed to obtain container engine info: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running? 
[black-reviewdog/lint      ] [DEBUG] evaluating expression 'success()'
[black-reviewdog/lint      ] [DEBUG] expression 'success()' evaluated to 'true'
[black-reviewdog/lint      ] 🚀  Start image=catthehacker/ubuntu:act-latest
[black-reviewdog/lint      ]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform= username= forcePull=true
[black-reviewdog/lint      ] [DEBUG]   🐳  docker pull catthehacker/ubuntu:act-latest
[black-reviewdog/lint      ] [DEBUG] pulling image 'docker.io/catthehacker/ubuntu:act-latest' ()
[Lint Black (official)/lint] [DEBUG] evaluating expression 'success()'
[Lint Black (official)/lint] [DEBUG] expression 'success()' evaluated to 'true'
[Lint Black (official)/lint] 🚀  Start image=catthehacker/ubuntu:act-latest
[Lint Black (official)/lint]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform= username= forcePull=true
[Lint Black (official)/lint] [DEBUG]   🐳  docker pull catthehacker/ubuntu:act-latest
[Lint Black (official)/lint] [DEBUG] pulling image 'docker.io/catthehacker/ubuntu:act-latest' ()
[Test Squire/Test Code     ] [DEBUG] evaluating expression 'success()'
[Test Squire/Test Code     ] [DEBUG] expression 'success()' evaluated to 'true'
[Test Squire/Test Code     ] [DEBUG] expression '${{ matrix.ubuntu }}' rewritten to 'format('{0}', matrix.ubuntu)'
[Test Squire/Test Code     ] [DEBUG] evaluating expression 'format('{0}', matrix.ubuntu)'
[Test Squire/Test Code     ] [DEBUG] expression 'format('{0}', matrix.ubuntu)' evaluated to '%!t(string=ubuntu-20.04)'
[Test Squire/Test Code     ] [DEBUG] expression '${{ matrix.ubuntu }}' rewritten to 'format('{0}', matrix.ubuntu)'
[Test Squire/Test Code     ] [DEBUG] evaluating expression 'format('{0}', matrix.ubuntu)'
[Test Squire/Test Code     ] [DEBUG] expression 'format('{0}', matrix.ubuntu)' evaluated to '%!t(string=ubuntu-20.04)'
[Test Squire/Test Code     ] [DEBUG] expression '${{ matrix.ubuntu }}' rewritten to 'format('{0}', matrix.ubuntu)'
[Test Squire/Test Code     ] [DEBUG] evaluating expression 'format('{0}', matrix.ubuntu)'
[Test Squire/Test Code     ] [DEBUG] expression 'format('{0}', matrix.ubuntu)' evaluated to '%!t(string=ubuntu-20.04)'
[Test Squire/Test Code     ] 🚀  Start image=catthehacker/ubuntu:act-20.04
[Test Squire/Test Code     ]   🐳  docker pull image=catthehacker/ubuntu:act-20.04 platform= username= forcePull=true
[Test Squire/Test Code     ] [DEBUG]   🐳  docker pull catthehacker/ubuntu:act-20.04
[Test Squire/Test Code     ] [DEBUG] pulling image 'docker.io/catthehacker/ubuntu:act-20.04' ()
Error: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

### Additional information

(On a repo with GitHub workflow actions, this one: https://github.com/LenaWil/Squire/commit/d038cc03cd0b075071a2ab001ed226c1b3953225 )

There are more workflows, and looking at the debug log it might run some of them? But without verbose mode it prints two error messages.

### PR DESCRIPTION
**Pull request by catthehacker**
Fixes #1658

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -118,6 +118,33 @@ func configLocations() []string {
...
+		if _, err := os.Lstat(os.ExpandEnv(p)); err == nil {
+			if strings.HasPrefix(p, `\\.\`) {
+				return "npipe://" + os.ExpandEnv(p), true
+			}
+			return "unix://" + os.ExpandEnv(p), true
+		}
+	}
+
+	return "", false
+}
```
Comment (wolfogre) [inline cmd/root.go:146]: How about:
```suggestion
// socketLocation returns socket path or false if not found any
func socketLocation() (string, bool) {
	if dockerHost, exists := os.LookupEnv("DOCKER_HOST"); exists {
		return dockerHost, true
	}

	for _, p := range commonSocketPaths {
		p = os.ExpandEnv(p)
		if info, err := os.Stat(p); err == nil {
			switch info.Mode().Type() {
			case os.ModeSocket:
				return "unix://" + p, true
			case os.ModeNamedPipe:
				return "npipe://" + p, true
			default:
				return p, true
			}
		}
	}

	return "", false
}
```

File: cmd/root.go [hunk 2]
```diff
@@ -118,6 +118,30 @@ func configLocations() []string {
...
+
+// returns socket path or false if not found any
+func socketLocation() (string, bool) {
+	if dockerHost, exists := os.LookupEnv("DOCKER_HOST"); exists {
+		return dockerHost, true
+	}
+
+	for _, path := range commonSocketPaths {
+		if _, err := os.Lstat(os.ExpandEnv(path)); err == nil {
```
[the remark this answers is not in the document]
Comment (wolfogre) [reply cmd/root.go]: I think we can check if the symlink is broken in the `bugReport`, but find a working one in `socketLocation`.

File: cmd/root.go [hunk 3]
```diff
@@ -150,22 +168,20 @@ func bugReport(ctx context.Context, version string) error {
...
 
 	report += sprintf("Docker host:", dockerHost)
 	report += fmt.Sprintln("Sockets found:")
 	for _, p := range commonSocketPaths {
-		if strings.HasPrefix(p, `$`) {
-			v := strings.Split(p, `/`)[0]
-			p = strings.Replace(p, v, os.Getenv(strings.TrimPrefix(v, `$`)), 1)
-		}
-		if _, err := os.Stat(p); err != nil {
+		if _, err := os.Stat(os.ExpandEnv(p)); err != nil {
```
Comment (wolfogre) [inline cmd/root.go:181]: How about:
```suggestion
		if _, err := os.Lstat(os.ExpandEnv(p)); err != nil {
```

File: cmd/root.go [hunk 4]
```diff
@@ -150,22 +168,20 @@ func bugReport(ctx context.Context, version string) error {
...
 			continue
-		} else {
-			report += fmt.Sprintf("\t%s\n", p)
 		}
+		report += fmt.Sprintf("\t%s\n", p)
```
Comment (wolfogre) [inline cmd/root.go:184]: Then:
```suggestion
		} else if _, err := os.Stat(os.ExpandEnv(p)); err != nil {
			report += fmt.Sprintf("\t%s(broken)\n", p)
		} else {
			report += fmt.Sprintf("\t%s\n", p)
		}
```

File: cmd/root.go [hunk 5]
```diff
@@ -118,6 +118,33 @@ func configLocations() []string {
 	}
 }
 
+var commonSocketPaths = []string{
```
Comment (LenaWil) [inline cmd/root.go:121]: @jpdrsn [claims](https://github.com/nektos/act/issues/1658#issuecomment-1517723861) `~/.rd/docker.sock` is also one of the possible locations.

**Review by wolfogre — APPROVED** [review]

**Pull request merged**